homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,324 @@
+    name = config.get(CONF_NAME)
+    entity_id = config.get(CONF_ENTITY_ID)
+    filters = []
+
+    for _filter in config[CONF_FILTERS]:
+        window_size = _filter.get(CONF_FILTER_WINDOW_SIZE)
+        precision = _filter.get(CONF_FILTER_PRECISION)
+
+        if _filter[CONF_FILTER_NAME] == FILTER_NAME_OUTLIER:
+            radius = _filter.get(CONF_FILTER_RADIUS)
+            filters.append(OutlierFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         radius=radius))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_LOWPASS:
+            time_constant = _filter.get(CONF_FILTER_TIME_CONSTANT)
+            filters.append(LowPassFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         time_constant=time_constant))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_THROTTLE:
+            filters.append(ThrottleFilter(window_size=window_size,
+                                          precision=precision,
+                                          entity=entity_id))
+
+    async_add_devices([SensorFilter(name, entity_id, filters)])
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, ICON)
+
+            self._state = new_state.state
+
+            if self._state == STATE_UNKNOWN:
+                return
+
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s, %s) -> %s", filt.name,
+                                  self._entity,
+                                  self._state,
+                                  "skip" if filt.skip else filtered_state)
+                    if filt.skip:
+                        return
+                except ValueError:
+                    _LOGGER.warning("Could not convert state: %s to number",
+                                    self._state)
+                finally:
+                    self._state = filtered_state
+                    filt.states.append(filtered_state)
```

Comment:
I am confused why this is necessary. Shouldn't the filter inside `filter_state` do this if it cares about the state? 

Also, since this is done inside `finally`, can't we just manage 1 list in the instance and pass it to all the filters ?

Reply:
states are per-filter, due to chaining the states list will be different for each filter